### PR TITLE
feat: missed-run catch-up on startup (#656)

### LINKED ISSUE #635: Catch-up approval UX: status bar badge + modal for missed scheduled runs
Parent: #589
Depends on: #633

Implements PR 2b of the background/scheduled tasks work.

## Scope

- [ ] On `onLayoutReady`, detect scheduled runs missed while Obsidian was closed (within a configurable window)
- [ ] Filter out tasks where `runIfMissed: false` — those are stale by definition and shouldn't appear in the catch-up list
- [ ] Collapse multiple missed runs of the same task into a single entry (no one wants 12 copies of the daily summary)
- [ ] Status bar `!` badge overlaid on the existing background icon when pending approvals exist
- [ ] Clicking the badge opens a modal listing pending runs with approve/skip buttons
- [ ] New setting: `autoRunCatchUp` (default `false`) — when enabled, skips the approval modal and runs missed tasks silently

## Non-goals

- Changing the core scheduler loop — that's #633

## Tests

- Missed-run detection window math
- `runIfMissed: false` filtering
- Collapse of duplicate missed runs
- `autoRunCatchUp` setting behavior

### PR DESCRIPTION
**Pull request by saheersk**
## Summary

Implements missed-run catch-up for scheduled tasks. When Obsidian is closed, scheduled tasks cannot run. Tasks with `runIfMissed: true` in their frontmatter are now detected on the next startup and surfaced in an approval modal so the user can decide whether to run or skip each one.

Fixes #635
Parent: #589
Depends on: #633

## Changes

- **`ScheduledTaskManager`** — add `detectMissedRuns()` (returns tasks with `runIfMissed: true` whose `nextRunAt` is in the past within a 7-day window) and `skipCatchUp()` (advances `nextRunAt` without running)
- **`PendingCatchUp` interface** — `{ task: ScheduledTask, missedAt: Date }` exported from `scheduled-task-manager.ts`
- **`BackgroundStatusBar`** — add `!` badge (red circle) that appears when there are pending catch-up approvals; click handler opens `CatchUpModal` instead of the regular background tasks modal when badge count > 0
- **`LifecycleService`** — `handleCatchUp()` called from `onLayoutReady()` after `initialize()` + `start()`; auto-runs silently when `autoRunCatchUp` is enabled, otherwise sets badge count
- **`CatchUpModal`** — new modal with per-row Run/Skip buttons and global Run all/Skip all; dismissing without acting leaves badge visible so the modal can be reopened
- **`autoRunCatchUp` setting** — new boolean in UI Settings ("Auto-run missed scheduled tasks on startup"); default `false`
- **`docs/guide/scheduled-tasks.md`** — updated `runIfMissed` description, added Catch-up Runs section covering modal behaviour, auto-run setting, and edge-case notes
- **Tests** — 6 new `detectMissedRuns` cases in `scheduled-task-manager.test.ts`; `ScheduledTaskManager` mock in `lifecycle-service.test.ts` updated with `detectMissedRuns`, `runNow`, `skipCatchUp`

## Screenshots / Screencast
Screenshot 2026-04-26 at 9 20 57 AM

**Missed Scheduled Runs modal (default behaviour):**
Modal appears on startup listing overdue tasks with Run / Skip per row and Run all / Skip all global actions. The `!` badge on the status bar persists if the modal is dismissed without acting.

Screenshot 2026-04-26 at 9 42 18 AM

**autoRunCatchUp enabled:**
No modal appears. Tasks are submitted silently as background tasks on startup (visible in the Background Tasks tab of the activity modal if still running).

## Manual Testing Performed

All scenarios tested by backdating `nextRunAt` in `scheduled-tasks-state.json` to yesterday, then fully quitting and restarting Obsidian (full quit required — "Reload app without saving" keeps the 60 s tick running which would fire and advance the date before startup catch-up runs):

| Scenario | Steps | Expected | Result |
|---|---|---|---|
| Modal appears on startup | `runIfMissed: true`, `nextRunAt` = yesterday, quit + restart | Missed Scheduled Runs modal opens automatically | ✅ |
| Skip (individual) | Click Skip on a row | Modal row removed; `nextRunAt` advances to next schedule; badge clears when last row handled | ✅ |
| Run (individual) | Click Run on a row | Task submitted as background task; `nextRunAt` advances; badge clears when last row handled | ✅ |
| Skip all | Click Skip all | All tasks skipped; modal closes; badge clears | ✅ |
| Run all | Click Run all | All tasks submitted; modal closes; badge clears | ✅ |
| Dismiss without acting | Press Escape / click ✕ | Modal closes; `!` badge remains on status bar | ✅ |
| Badge click reopens modal | After dismiss, click `!` badge | Catch-up modal reopens (not regular background tasks modal) | ✅ |
| `autoRunCatchUp = true` | Enable setting, backdated date, quit + restart | No modal, no badge; task submitted silently; visible in Background Tasks modal | ✅ |
| Task paused — excluded | `pausedDueToErrors: true` | Task not shown in catch-up modal | ✅ (covered by unit test) |
| Outside 7-day window — excluded | `nextRunAt` > 7 days ago | Task not shown in catch-up modal | ✅ (covered by unit test) |
| `runIfMissed: false` — excluded | Default value | Task not shown in catch-up modal | ✅ (covered by unit test) |

### Required

- [x] I have read and agree to the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have read and agree to the [AI Policy](../AI_POLICY.md)
- [x] This PR is linked to an approved issue where the approach was discussed with a maintainer
- [x] All CI checks pass (`npm test`, `npm run build`, `npm run format-check`)
- [x] I have tested this change on Desktop
- [x] I have verified this change does not break Mobile (or includes appropriate platform guards)
  - Tested via `app.emulateMobile(true)` in DevTools — mobile layout renders correctly, modal is usable
  - **Known limitation:** Obsidian hides the status bar on mobile, so the `!` badge is not visible. The approval modal still auto-opens on startup via `handleCatchUp()`. If the user dismisses without acting, there is currently no way to reopen the modal on mobile (no badge to click). A follow-up could add a command palette entry (`Open Missed Runs`) to cover this gap.
- [x] Documentation has been updated (if applicable)
- [x] I understand that I must address all review comments from CodeRabbit and maintainers, or this PR may be closed

### AI-Generated Code

- [x] This PR includes AI-generated or AI-assisted code
- [x] AI tool(s) used: Claude (claude-sonnet-4-6 via Claude Code)
- [x] I have reviewed and understand all AI-generated code in this PR

## Summary by CodeRabbit

* **New Features**
  * Catch-up workflow for missed scheduled tasks with a “Missed Scheduled Runs” modal, per-task Run/Skip and bulk Run all/Skip all actions.
  * New toggle to auto-run missed tasks on startup and a status-bar badge showing pending catch-up approvals.

* **Documentation**
  * Guide updated to explain missed-run handling, modal behavior, auto-run option, eligibility rules, and timing constraints.

* **Style**
  * Added UI styles for catch-up modal and badge.

* **Tests**
  * Added tests for catch-up detection and lifecycle handling.

### CODE REVIEW

**Comment by allenhutchison** [comment]
Hey @saheersk — thanks for the great work on this! 👋

Heads-up: I pushed a few commits directly to your branch. Wanted to give you context on why and what changed so nothing feels surprising.

## Why we jumped in

The vitest migration (#699) landed on master after you'd already opened this PR, and #696/#702 followed quickly after that. That left your branch conflicting on the test files even though your code itself was fine. Since you'd been responsive to review and the feature was in great shape, it seemed friendlier to do the rebase + tiny cleanups for you rather than ask you to redo all the test plumbing.

To be clear: **the catch-up design and implementation here are excellent.** The state machine (`catchUpPending` Set + `reserveForCatchUp` + tick-loop skip), the auto-run vs. badge-and-modal split, the 7-day stale window, the "skip" path that advances `nextRunAt` so the task isn't re-detected — all really clean. The CodeRabbit pass you did before us already addressed most of what would normally come back in review.

## What I changed and why

**1. Rebased onto master.** Test files now use `vi.*` instead of `jest.*`. While resolving conflicts in `test/services/lifecycle-service.test.ts` I also added the new methods to the `ScheduledTaskManager` mock (`detectMissedRuns`, `runNow`, `skipCatchUp`, `reserveForCatchUp`) so future tests can exercise them.

**2. Status bar visibility bug** (the one CodeRabbit flagged in `background-status-bar.ts`). The hide guard `if (runningCount === 0 && ragStatus === 'disabled')` would hide the entire status bar item even when `_pendingCatchUpCount > 0`, making the `!` badge unreachable for users who don't run RAG. I added `&& this._pendingCatchUpCount === 0` to the guard.

While I was in there I also added a small UX touch: when the bar is visible *only* because of pending catch-ups (no tasks, RAG disabled), it now shows a clock icon instead of falling through to the database icon — that way the badge isn't sitting next to a misleading icon. Tooltip gets a "N missed scheduled run(s) — click to review" line in the same case. Live-tested on my vault: bar shows clock + badge with RAG off, hides cleanly once you Skip the run.

**3. Lifecycle behavioral tests.** CodeRabbit noted (correctly) that the `detectMissedRuns`/`runNow` mock entries I added were just shape, no actual behavior assertions. Added four tests in `lifecycle-service.test.ts`:
- `autoRunCatchUp: true` → `runNow` called once per missed task, badge not set, slugs not reserved
- `autoRunCatchUp: false` → slugs reserved, badge count set, `runNow` not called
- empty `detectMissedRuns` → no-op
- single `runNow` rejection → loop continues to subsequent tasks

## CodeRabbit findings I deliberately didn't act on

A few of CodeRabbit's other findings were either already addressed in your latest commit or based on a misread of the codebase. For the record:

- **`InstanceType<typeof ObsidianGemini>` typing nit** — this codebase actually uses `private plugin: ObsidianGemini` in every modal (`background-tasks-modal`, `scheduled-tasks-modal`, etc.). Yours matches the convention; CodeRabbit's "learning" was inverted.
- **`var(--font-smaller)` token** — your CSS uses `var(--font-ui-smaller)`, which matches 30+ other usages in `styles.css`. CodeRabbit hallucinated the wrong token name.
- **"`handleCatchUp` requires `backgroundStatusBar`"** — wrong; you use `?.` on it.
- **"Tick can submit overdue tasks during/after handleCatchUp"** — already covered by your `catchUpPending` Set + `reserveForCatchUp`.
- **"Self-correct stale badge in click handler"** — already there at the line `// detectMissedRuns returned empty — stale badge; self-correct`.
- **"Modal handlers need try/catch + button-disable"** — already implemented.

## Live testing

Reloaded the plugin in my vault and exercised both flows:
- Manual: created a task with `runIfMissed: true` and a `nextRunAt` 1h ago → reload → status bar visible with clock icon and `!` badge → click opens modal showing "missed 1h ago" → Skip advances `nextRunAt` to tomorrow, badge clears, status bar hides.
- Auto: same setup with `autoRunCatchUp: true` → reload → log shows `[LifecycleService] Auto catch-up: submitted "catchup-test"` → BackgroundTaskManager runs and completes it, badge stays at 0.

All 1476 tests passing, format/build clean.

Really appreciate you tackling #635 — the catch-up UX is a meaningful quality-of-life improvement. Hope the bit of help moves it across the line a little faster. If anything in the changes doesn't sit right with you, push back; happy to revisit. 🙏

File: docs/guide/scheduled-tasks.md [hunk 1]
```diff
@@ -102,9 +102,31 @@ To resume a paused task:
...
+
+When Obsidian is closed, scheduled tasks cannot run. Set `runIfMissed: true` in a task's frontmatter to have it caught up automatically the next time Obsidian starts.
+
+On startup, the plugin compares each task's `nextRunAt` against the current time. Any task with `runIfMissed: true` that is overdue (and not paused) is treated as a missed run. By default, a **Missed Scheduled Runs** modal appears listing each missed task with **Run** and **Skip** buttons:
+
+- **Run** — submits the task immediately as a background task
+- **Skip** — advances the schedule without running
+- **Run all / Skip all** — bulk actions for all listed tasks
+
+Dismissing the modal (Escape or ✕) leaves the `!` badge on the status bar so you can reopen it later by clicking the badge.
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply docs/guide/scheduled-tasks.md]: Good catch — fixed in b0394d3. The docs now describe the badge-first flow accurately: `handleCatchUp()` sets the badge count and the user clicks to open the modal. The auto-run setting is correctly described as silent execution (no modal involvement) already. Thanks!
